fitlins/workflows/base.py

```diff
@@ -150,160 +76,134 @@ def join_dict(base_dict, dict_list):
-        (collate_second_level, ds_l2_contrast_maps, [('contrast_maps', 'in_file'),
-                                                     ('contrast_metadata', 'entities')]),
-
-        (select_l1_entities, ds_design, [('out', 'entities')]),
-        (plot_design, ds_design, [('figure', 'in_file')]),
-
-        # (select_l1_entities, ds_corr, [('out', 'entities')]),
-        # (plot_corr, ds_corr, [('figure', 'in_file')]),
-
 
-        # (select_l1_entities, ds_l1_contrasts, [('out', 'entities')]),
-        # (plot_l1_contrast_matrix, ds_l1_contrasts, [('figure', 'in_file')]),
-
-        (collate_first_level, ds_l1_contrast_plots, [('contrast_metadata', 'entities')]),
-        (plot_l1_contrasts, ds_l1_contrast_plots, [('figure', 'in_file')]),
-
-        # (plot_l2_contrast_matrix, ds_l2_contrasts, [('figure', 'in_file')]),
-
-        (collate_second_level, ds_l2_contrast_plots, [('contrast_metadata', 'entities')]),
-        (plot_l2_contrasts, ds_l2_contrast_plots, [('figure', 'in_file')]),
-        ])
+        ds_contrast_maps = pe.Node(
+            BIDSDataSink(base_directory=out_dir,
+                         path_patterns=contrast_pattern),
+            run_without_submitting=True,
+            name='ds_{}_contrast_maps'.format(step))
+
+        ds_contrast_plots = pe.Node(
+            BIDSDataSink(base_directory=out_dir,
+                         path_patterns=contrast_plot_pattern),
+            run_without_submitting=True,
+            name='ds_{}_contrast_plots'.format(step))
+
+        if step == 'run':
+            # Select preprocessed BOLD series to analyze
+            getter = pe.Node(
+                BIDSSelect(
+                    bids_dir=bids_dir, derivatives=derivatives,
+                    selectors={
+                        'type': 'preproc', 'suffix': 'bold', 'space': space}),
+                name='getter')
+
+            # Run first level model
+            model = pe.MapNode(
+                FirstLevelModel(),
+                iterfield=['session_info', 'contrast_info', 'bold_file', 'mask_file'],
+                name='{}_model'.format(step))
+            models.append(model)
+
+            def join_dict(base_dict, dict_list):
+                return [{**base_dict, **iter_dict} for iter_dict in dict_list]
+
+            # Accumulate metadata
+            l1_metadata = pe.MapNode(
+                niu.Function(function=join_dict),
+                iterfield=['base_dict', 'dict_list'],
+                name='l1_metadata')
+
+            plot_design = pe.MapNode(
+                DesignPlot(image_type='svg'),
+                iterfield='data',
+                name='plot_design')
+
+            # Images
+            ds_design = pe.MapNode(
+                BIDSDataSink(base_directory=out_dir, fixed_entities={'type': 'design'},
+                             path_patterns=image_pattern),
+                iterfield=['entities', 'in_file'],
+                run_without_submitting=True,
+                name='ds_design')
```

Comment:
All of these things that apply only to L1 should not happen in the loop.

Reply:
that makes sense